### PR TITLE
feat(Translations): Translations for YouTube and Twitch

### PR DESCRIPTION
**Pull request by LisoUseInAIKyrios**
Removed usage of `StringResource`, and moved all strings into Strings.xml files which allows easy localization.

Added translations for a few random languages (machine translated to show it works.  Preferably should be replaced with Crowdin or some other human translated source)

Translation files are bundled automatically, just add or remove the translated Strings.xml files to either `youtube/translation` or `twitch/translation`.

Hebrew:
hebrew settings
hebrew sponsorblock

French (_ooh là là_):
french sponsorblock

French Twitch:
french twitch

[integration changes](https://github.com/revanced/revanced-integrations/pull/420)

### CODE REVIEW

**Comment by 0xrxL** [comment]
Crowdin is a better solution imho.

Better to translator too (with its UX).

**Comment by 0xrxL** [comment]
@oSumAtrIX Do you think Crowdin it's a good solution ReVanced translations?

**Comment by oSumAtrIX** [comment]
> Crowdin is a better solution imho.
> 
> Better to translator too (with its UX).

This PR is necessary for Crowdin to work.

**Comment by oSumAtrIX** [comment]
This issue is partially related to AAPT2. It may work with AAPT1 and will be fixed once we stop using aapt

**Comment by oSumAtrIX** [comment]
A crowdin project can be set up for this, but I am not sure, how we would set this up properly for this feature branch. Once merged to dev or main, translations can be submitted

**Comment by Axelen123** [comment]
> This issue is partially related to AAPT2. It may work with AAPT1 and will be fixed once we stop using aapt

I am not sure how we should deal with this after the arsclib branch. That branch no longer uses `strings.xml` because the encoder for it wasnt available at the time.

We have three options:
- Use ARSCLibs encoder which was recently made public, but doesnt handle escapes
- Write our own strings.xml encoder
- Use a different format

**Comment by 0xrxL** [comment]
@oSumAtrIX What's still missing on this PR before it can be merged?

**Comment by oSumAtrIX** [comment]
A review, it's a big PR

File: src/main/kotlin/app/revanced/patches/youtube/misc/settings/bytecode/patch/YouTubeSettingsPatch.kt
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/youtube/misc/settings/bytecode/patch/YouTubeSettingsPatch.kt:1]: While I understand the motivation behind renaming this class of not accidentally choosing the "wrong" `SettingsPatch` class, the package name already includes the application the patch is made for, so adding the app to the class name of the patch is a duplicate in this regard. This change may be reconsidered

**Review by oSumAtrIX — CHANGES_REQUESTED** [review]
The only problem with this PR is that it breaks the separation of patches. Now a single file manages multiple patches. This means the apps will contain traces of patches that should not be there, to begin with. A similar issue is seen with integrations currently, which will eventually be solved once integrations are coupled with their respective patches. Is it possible to separate the files so that traces of patches I did not apply do not appear in the patched app? I was following the `host` convention so far for that reason. Some patches delivered resources via a "host". The respective patch then uses the host to add the patch's respective resources to the app. This may be a huge task to refactor for this PR which is why if it is, do not hesitate to suggest a solution that can be approached together.

**Comment by optimatrix0** [comment]
I'm looking forward to being implemented

**Comment by sssh7710** [comment]
Hasn't it been officially released yet?

**Comment by oSumAtrIX** [comment]
The PR status is open.

**Comment by oSumAtrIX** [comment]
@LisoUseInAIKyrios The strategy you follow here is to include a "Translations" patch that the user can select to include.  What happens if they do not? The resource files, thus the strings will never get merged.

**Comment by oSumAtrIX** [comment]
I am hesitant about this PR (sorry for the long time) because the PR breaks the modular nature of patches by introducing a file that is not separate from each patch.

But I can understand the motivation for using a single file, though. If every patch had its own translations, that would be an immense amount of files.

You suggest that each patch has its own strings.xml file, but each would have its English language file. The goal would be that you can exclude the translations patch but still have English translations. This makes sense to me.

Some changes I'd like to suggest for this PR:

1. Currently, resources are organized into patches: 
image.
   This shows as impractical as we can see with the translations patch. Multiple apps have the same patch, but we can not create one patch path "translations" for all of them. Because the source packages are divided by apps, I suggest to also change the current convention of dividing resources by their patch in addition to dividing them by their app. This means, that youtube patches resources should be moved into the "youtube" resource path, and twitch patches resources, to a "twitch" resource path with this PR.

2. The settings patches have been renamed to their respective target app (SettingsPatch -> TwitchSettingsPatch). Despite the obvious motivation for this, I suggest not renaming them. Patches are currently organized into packages for the respective app, which already separates SettingsPatch A from SettingsPatch B. No other patch is prefixed with the target app and patches already have constraints to the target app via the @Compatibility annotation. Therefor, despite the small annoyance of the IDE showing the same name, I still suggest remaining consistent in naming the patches and following the current convention of separating the patches by package. (PS: We should use the full package name of the apps the patches target in favour of correctness: music -> com.google.android.apps.youtube.music)

3. Currently, every app has received a "translations" patch. To reduce duplicate code, abstraction can be used. A shared abstract patch can be used. You can refer to the Reddit client patches, for example.

**Comment by oSumAtrIX** [comment]
Can you reproduce by moving a single string to a single file?

**Comment by oSumAtrIX** [comment]
AAPT not printing any message might originate from the way the process error stream is read by Androlib. I am fairly sure that can be outputted to the logger.

**Comment by oSumAtrIX** [comment]
Why is Gradle needed?

**Comment by Ushie** [comment]
> Still not sure if Crowdin can import multiple string files into a single project (and if it automatically includes any new patch string files added later)

Multiple source files can be added, and multiple repositories even

**Comment by Ushie** [comment]
The current ReVanced Manager Crowdin is considered unused, it's waiting for the Compose branch of ReVanced Manager to be merged before it's restarted

I was planning to leave it untouched until then, but if you want I can clear everything and prepare it for ReVanced Patches sometime

File: src/main/kotlin/app/revanced/util/resources/ResourceUtils.kt
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/util/resources/ResourceUtils.kt:118]: I have yet to understand the reason for this entirely.
The XML files are in the resources folder, which Crowdin can pick up, is there any reason why a Gradle task is necessary

File: src/main/kotlin/app/revanced/util/resources/ResourceUtils.kt
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/util/resources/ResourceUtils.kt:138]:
```suggestion
        rootElement.appendChild(rootElement.ownerDocument.createComment("This file is generated by a Gradle task and should never be manually edited"))
```

File: src/main/kotlin/app/revanced/util/resources/ResourceUtils.kt
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/util/resources/ResourceUtils.kt:87]: What we refer to "resource paths" or "resource directories" does not exist. Zip files do not have the semantic of folders and instead every folder we see in zip files is actually parsed from ZipEntry name. This means that every zip or jar file only has zip files and no folder structure. Because of that, jar resource file names should always be known at compile time whereas this method tries to retrieve the resource files at runtime dynamically. Instead, combining with my other comment regarding creating `values-<countrycode>` folders, and patches depending on the translations patch, the patch could iterate over each patch it registered, and try to load each resource file for every country code it knows (a list of country codes can dynamically be retrieved using `Locale.getISOCountries()`). The patch would try reading the resource, if it does not exist, try reading the resource file for the next country code, for each patch. This approach would spare relying on the approach seen in this method.

File: src/main/kotlin/app/revanced/util/resources/ResourceUtils.kt
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/util/resources/ResourceUtils.kt:90]: j

File: build.gradle.kts
Comment (oSumAtrIX) [reply build.gradle.kts]: > Because the task only merges the individual patch string files, which are English.

I believe merging is not necessary. Crowdin can accept a wildcard for source files to translate.

If the only reason for merging was to support Crowdin this can safely be removed.

File: src/main/kotlin/app/revanced/util/resources/ResourceUtils.kt
Comment (oSumAtrIX) [reply src/main/kotlin/app/revanced/util/resources/ResourceUtils.kt:39]: Since AAPT is failing, the correct way to handle this case is to log the error output stream when executing the AAPT process. This way, any invalid resources would give proper feedback at compile time. But implementing that is not related to this PR which is why I was suggesting to remove this code, as the correct place to handle it would be outside the PR.

Comment (oSumAtrIX) [reply src/main/kotlin/app/revanced/util/resources/ResourceUtils.kt:33]: > include the string but strip the quotes, and that cannot possible be what the developer wanted

Would you classify this as a bug in AAPT? I believe it may have good reason the way it is.

File: src/main/kotlin/app/revanced/patches/youtube/layout/spoofappversion/SpoofAppVersionPatch.kt
Comment (oSumAtrIX) [reply src/main/kotlin/app/revanced/patches/youtube/layout/spoofappversion/SpoofAppVersionPatch.kt:75]: > is typos will be treated at literals

Another option would be to simply input the full string value. This means, if you want to use a literal, you'd enter a literal value, but if you want to use a reference, you'd input `@string/value`. This prevents having to create boilerplate classes, manually specifying if the values are literals and allows for each entry to be a reference or literal.

Comment (oSumAtrIX) [reply src/main/kotlin/app/revanced/patches/youtube/layout/spoofappversion/SpoofAppVersionPatch.kt:75]: > That would mean boilerplate @string/ prefix is needed. 

Yes. A certain amount of boilerplate will be necessary. May it be in the form of abstraction or treating the attribute value literally. It would certainly get rid of the parameter to specify if the inputs are literals as well as allow each value to be literal or reference though. The added boilerplate is minimal though and straight forward to use whereas I had to first understand why a parameter to specify the values as literals exists.

>  And it's not consistent with where keys are used everywhere else without declaring it's for a string resource.

Keys are different from values, the parameters ask for a key specifically, but here we ask for values, or am I understanding you incorrectly?

Comment (oSumAtrIX) [reply src/main/kotlin/app/revanced/patches/youtube/layout/spoofappversion/SpoofAppVersionPatch.kt:75]: Yeah, that way my initial solution above too but it would require the user to wrap every string in one, which is cumbersome, I'd rather leave it as it is right now, otherwise, if you want, we can use the abstraction

File: src/main/resources/youtube/settings/host/values/AlbumCards.xml
Comment (oSumAtrIX) [reply src/main/resources/youtube/settings/host/values/AlbumCards.xml:2]: I understand your concerns regarding too many files and the maintenance aspect of it.

There's a structure of the "settings" resource directory hosting any strings for patches. It doesn't sound like a patch has to depend on `SettingsPatch` to merge some strings. An example would be introducing a patch that shows a toast but does not add in app settings and thus does not require the `Settings` patch.

My main motivation for separating the files is that each patch will be separated from each other, preserving the modular structure of patches. As of now, the translation patch hosts all the translations. While maintaining thousands of files would be problematic, these files would be generated automatically and never supposed to touch. Each patch would have a folder that would host the translations. Here, the folder structure is unnecessary; Crowdin could commit the translated files directly into the respective patches' translation folder, whereas the translation patch would pick them up and merge them to be added to the app. This would prevent having to create hundreds of folders, and instead, each patch would have its own translations under `<patchname>/translations`. Additionally, this structure preserves the modular structure of patches. If I want to remove a patch, for example, I can eliminate the resource directory for the patch and do not have to worry about first committing them, letting Crowdin register the removed strings and remove them from the translations.

The workflow would look the following way:

1. Create the resource directory for the translations (`<patchname>/translations>`)
2. Create the English `strings-en_US.xml` file
3. Commit and let Crowdin translate the files and commit the translations to `<patchname>/translations>`
4. The patches depend on the translations patch, which would pick up the translated files and create the `values` directories accordingly to write the strings file.

Let me know what you think.

Comment (oSumAtrIX) [reply src/main/resources/youtube/settings/host/values/AlbumCards.xml:2]: I believe it should be possible for Crowding to write translations to the same path of a source file. This will work for any number of source files. @Ushie can you confirm?

Comment (Ushie) [reply src/main/resources/youtube/settings/host/values/AlbumCards.xml:2]: Can you TL;DR? to the same directory? yes

Comment (oSumAtrIX) [reply src/main/resources/youtube/settings/host/values/AlbumCards.xml:2]: Basically, if I have multiple source strings files, each in separate directories, can Crowdin create translations for each of the source files in each of their directories?

Comment (Ushie) [reply src/main/resources/youtube/settings/host/values/AlbumCards.xml:2]: Yes, each imported source can have it's export directory managed individually

Comment (oSumAtrIX) [reply src/main/resources/youtube/settings/host/values/AlbumCards.xml:2]: @LisoUseInAIKyrios If there isn't anything else, we can start orchestrating anything that needs manual refactoring that I can help with.

Comment (oSumAtrIX) [reply src/main/resources/youtube/settings/host/values/AlbumCards.xml]: > It's a lot of clutter when each patch patch has 20 language files, but only one file can be edited

Yes. Each patch has its own language files. I primarily took that into account to preserve the modularity of patches. Because the files are automatically generated, it is fine that there are many language files. Maybe Crowdin has a feature that allows you to write all translations into a single file (per patch). This way, if we have 100 translatable patches, we will have 100 additional translated string files. The translation patch could then parse these files and merge all 100 translation files into one by mapping them to the corresponding language. But the prerequisite is that Crowdin supports that. If not, an intermediate branch Crowdin pushes to can trigger CI to apply the transformations. Otherwise, dealing with all the translation files is fine too, because they are generated automatically.

> The translation patch is also more clunky, because it has to merge 1,000 files and it needs to figure out which patches were included and which were not. 

It would have to iterate over each patch that it depends on, get all translated files, and map their translated strings to their respective language.

> if the root contains only a single subdirectory

You can disable flattening the path:

image

PS: You can use IntelliJ IDEA for all ReVanced projects, including integrations by importing it as a Gradle module, this gets rid of the need to use Android Studio.

Comment (oSumAtrIX) [reply src/main/resources/youtube/settings/host/values/AlbumCards.xml]: Please let me know if you still have any concerns.

Comment (oSumAtrIX) [reply src/main/resources/youtube/settings/host/values/AlbumCards.xml]: Each patch would have a resource directory attributed and named after itself, meaning, that you would be able to locate the string resource by querying the directory names with similar effort. Considering though, that the strings would be saved in the translations patch resource folder, I agree that it makes sense to skip this overhead and name the string files after the patch. I suggest to prefix or suffix it with "string" for example: "HideAds_strings.xml", "strings_HideAds.xml".

Regarding your previous response I still need some time to reply to it.

**Comment by oSumAtrIX** [comment]
Hey, just letting you know that we have renamed repositories. Please close this PR and re-open it [here](https://github.com/ReVanced/revanced-patches/pulls). PS: No additional changes are needed to your remotes.

**Comment by LisoUseInAIKyrios** [comment]
Pull request moved to https://github.com/ReVanced/revanced-patches/pull/2440

**Pull request closed without merging**
